pymecht/MatModel.py

```diff
@@ -1044,4 +1044,136 @@ def _energy_stress(self,A,B,mean_theta,sigma,aniso_fraction,**extra_args):
+                print("Invalid format. Please try again.")
+        success = False
+        while success == False:
+            try:
+                if _low_bound!='':
+                    self.param_low_bd = eval('dict('+_low_bound+')')
+                else:
+                    self.param_low_bd   = eval('dict('+input("Please enter lower bound for parameters in the form PARAM1=VAL1, PARAM2=VAL2, ..., PARAMN=VALN: ")+')')#dict(c1=0.0001,c2=0.,c3=0.,c4=0.)
+                if len(self.param_low_bd) == 0:
+                    print("Warning: Lower bound not set for parameters")
+                success = True
+            except:
+                print("Invalid format. Please try again.")
+        success = False
+        while success == False:
+            try:
+                if _up_bound!='':
+                    self.param_up_bd = eval('dict('+_up_bound+')')
+                else:
+                    self.param_up_bd    = eval('dict('+input("Please enter upper bound for parameters in the form PARAM1=VAL1, PARAM2=VAL2, ..., PARAMN=VALN: ")+')')#dict(c1=100.,c2=100.,c3=100.,c4=100.)
+                if len(self.param_up_bd) == 0:
+                    print("Warning: Upper bound not set for parameters")
+                success = True
+            except:
+                print("Invalid format. Please try again.")
+        
+        # Get the user's strain energy density function (SEDF)
+        if _W!='':
+            self.energy_form = _W
+        else:
+            self.energy_form = input("Please enter form of the SEDF: ")#"c1*(I1-3)+c2*(I1-3)**2+c3*(I1-3)**3+c4*(I1-3)**4"
+        # Since there are an arbitrary number of arbitrarily-named parameters,
+        # we get these from the definition of the initial guess.
+        self.param_names = [i for i in self.param_default]
+        
+        # Need sympy for symbolic differentiation
+        # Derivatives calculated here to avoid runtime being called in loop
+        import sympy as sp
+        
+        # Define invariants as symbols
+        I1,I2,I3,I4 = sp.symbols('I1 I2 I3 I4')
+        
+        # Construct a string of the symbols to define as such without needing
+        # numerical values
+        param_names_string = ""
+        for name in self.param_names:
+            param_names_string += name + " "
+        
+        other_symbols = sp.symbols(param_names_string)
+        
+        SEDF = sp.expand(self.energy_form)
+        
+        # Take symbolic derivatives
+        dSEDFdI =   sp.diff(SEDF,I1), \
+                    sp.diff(SEDF,I2), \
+                    sp.diff(SEDF,I3), \
+                    sp.diff(SEDF,I4)
+        
+        # Store derivatives as strings
+        dSEDFdI = [str(sp.powdenest(sp.factor_terms(sp.powsimp(i)))) for i in dSEDFdI]
+        
+        # Format derivatives for consistent syntax
+        dSEDFdI = self.replaceSympySyntax(dSEDFdI)
+        
+        # Update attribute of class.
+        self.denergy_formdI = dSEDFdI
+
```

Comment:
Here, we should add a check here if the expression contains I4 term since that would need a fiber direction (see other model functions). Something like 
```
if 'I4' in self.energy_form:
   self.I4term = True
else:
   self.I4term = False
```

Reply:
I elected for `self.I4term = ('I4' in self.energy_form)`.